fuel-vm/src/tests/alu.rs

```diff
@@ -275,491 +70,375 @@ fn reserved_register() {
+#[test_case(op::modi, u64::MAX, 2 => AluResultForFlags::invariant_ok(u64::MAX % 2))]
+#[test_case(op::expi, 0, 0 => AluResultForFlags::invariant_ok(1))]
+#[test_case(op::expi, 2, 3 => AluResultForFlags::invariant_ok(8))]
+#[test_case(op::expi, 3, 3 => AluResultForFlags::invariant_ok(27))]
+#[test_case(op::expi, 1, 100 => AluResultForFlags::invariant_ok(1))]
+#[test_case(op::expi, u64::MAX, 1 => AluResultForFlags::invariant_ok(u64::MAX))]
+#[test_case(op::expi, u32::MAX as u64, 2 => AluResultForFlags::invariant_ok((u32::MAX as u64).pow(2)))]
+#[test_case(op::expi, u64::MAX, 2 => AluResultForFlags::wrapping_fixed())]
+#[test_case(op::slli, 0, 0 => AluResultForFlags::invariant_ok(0))]
+#[test_case(op::slli, 1, 0 => AluResultForFlags::invariant_ok(1))]
+#[test_case(op::slli, 1, 1 => AluResultForFlags::invariant_ok(1 << 1))]
+#[test_case(op::slli, u64::MAX, 4 => AluResultForFlags::invariant_ok(u64::MAX << 4))]
+#[test_case(op::slli, u64::MAX, 100 => AluResultForFlags::invariant_ok(0))]
+#[test_case(op::srli, 0, 0 => AluResultForFlags::invariant_ok(0))]
+#[test_case(op::srli, 1, 0 => AluResultForFlags::invariant_ok(1))]
+#[test_case(op::srli, 1, 1 => AluResultForFlags::invariant_ok(0))]
+#[test_case(op::srli, 0b10, 1 => AluResultForFlags::invariant_ok(1))]
+#[test_case(op::srli, u64::MAX, 4 => AluResultForFlags::invariant_ok(u64::MAX >> 4))]
+#[test_case(op::srli, u64::MAX, 100 => AluResultForFlags::invariant_ok(0))]
+#[test_case(op::andi, 0b1111, 0b1111 => AluResultForFlags::invariant_ok(0b1111))]
+#[test_case(op::andi, 0b1101, 0b0111 => AluResultForFlags::invariant_ok(0b0101))]
+#[test_case(op::andi, u64::MAX, 0 => AluResultForFlags::invariant_ok(0))]
+#[test_case(op::andi, u64::MAX, 1 => AluResultForFlags::invariant_ok(1))]
+#[test_case(op::ori, 0b1111, 0b1111 => AluResultForFlags::invariant_ok(0b1111))]
+#[test_case(op::ori, 0b1101, 0b0111 => AluResultForFlags::invariant_ok(0b1111))]
+#[test_case(op::ori, 0b1101, 0b0101 => AluResultForFlags::invariant_ok(0b1101))]
+#[test_case(op::ori, u64::MAX, 0 => AluResultForFlags::invariant_ok(u64::MAX))]
+#[test_case(op::ori, u64::MAX, 1 => AluResultForFlags::invariant_ok(u64::MAX))]
+#[test_case(op::xori, 0b1111, 0b1111 => AluResultForFlags::invariant_ok(0))]
+#[test_case(op::xori, 0b1101, 0b0111 => AluResultForFlags::invariant_ok(0b1010))]
+#[test_case(op::xori, 0b1101, 0b0101 => AluResultForFlags::invariant_ok(0b1000))]
+#[test_case(op::xori, u64::MAX, 0 => AluResultForFlags::invariant_ok(u64::MAX))]
+#[test_case(op::xori, u64::MAX, 1 => AluResultForFlags::invariant_ok(u64::MAX ^ 1))]
+fn test_binary_op_reg_imm(
+    op: fn(u8, u8, u16) -> Instruction,
+    lhs: Word,
+    rhs: u16,
+) -> AluResultForFlags {
+    let run = |flags| run_alu_op(op(0x20, 0x10, rhs), vec![lhs], flags);
+    AluResultForFlags {
+        normal: run(Flags::empty()),
+        wrapping: run(Flags::WRAPPING),
+        unsafemath: run(Flags::UNSAFEMATH),
+        wrapping_unsafemath: run(Flags::WRAPPING | Flags::UNSAFEMATH),
```

Comment:
Why isn't this `&` instead of `|`? Am I missing something? (same thing below)

Reply:
Thse are bitflags. `|` combines them. imagine unsafemath = 1 (`0b01`), wrapping = 2 (`0b10`). Now `0b01 & 0b10 == 0b00` and `0b01 | 0b10 == 0b11`

Reply:
Cool. I see now.